test/SlowTests/Issues/RavenDB-11571.cs

```diff
@@ -0,0 +1,178 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Exceptions;
+using Raven.Client.ServerWide;
+using Raven.Server.Utils;
+using Raven.Tests.Core.Utils.Entities;
+using Tests.Infrastructure;
+using Xunit;
+
+namespace SlowTests.Issues
+{
+    public class RavenDB_11571 : ClusterTestBase
+    {
+        [Fact]
+        public void CounterStorage_Increment_ShouldCheckForOverflow()
+        {
+            using (var store = GetDocumentStore())
+            {
+                using (var session = store.OpenSession())
+                {
+                    session.Store(new User
+                    {
+                        Name = "Aviv"
+                    }, "users/1");
+                    session.CountersFor("users/1").Increment("Downloads", long.MaxValue);
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var documentCounters = session.CountersFor("users/1");
+                    var val = documentCounters.Get("Downloads");
+                    Assert.Equal(long.MaxValue, val);
+
+                    documentCounters.Increment("Downloads");
+                    var e = Assert.Throws<RavenException>(() => session.SaveChanges());
+                    Assert.Contains("Arithmetic operation resulted in an overflow", e.Message);
```

Comment:
We need to tell on what counter / document this happened.

Reply:
@aviv86 Let's create a CounterOverflowException for that then. We can push there document id, counter name, old value and new value and OverflowException as inner.

Reply:
E.g. `Could not increment counter '{name}' from document '{documentId}' with value '{value}' by '{increment}'. Arithmetic operation resulted in an overflow`

Reply:
done